blog/2020/05/invalidations.md

```diff
@@ -0,0 +1,734 @@
+            MethodInstance for collect_project!(::Pkg.Types.Context, ::Pkg.Types.PackageSpec, ::String, ::Dict{Base.UUID,Array{Pkg.Types.PackageSpec,1}})
+             MethodInstance for collect_fixed!(::Pkg.Types.Context, ::Array{Pkg.Types.PackageSpec,1}, ::Dict{Base.UUID,String})
+             ⋮
+```
+
+This indicates that this invalidation path resulted from a call to `maximum` from a function in `Pkg`, `set_maximum_version_registry!`.  A little digging reveals that it is defined as
+
+```
+function set_maximum_version_registry!(ctx::Context, pkg::PackageSpec)
+    pkgversions = Set{VersionNumber}()
+    for path in registered_paths(ctx, pkg.uuid)
+        pathvers = keys(load_versions(ctx, path; include_yanked=false))
+        union!(pkgversions, pathvers)
+    end
+    if length(pkgversions) == 0
+        pkg.version = VersionNumber(0)
+    else
+        max_version = maximum(pkgversions)
+        pkg.version = VersionNumber(max_version.major, max_version.minor, max_version.patch, max_version.prerelease, ("",))
+    end
+end
+```
+
+From that error above, we know that this invalidation is produced by an error-path triggered by trying to reduce over an empty collection.
+Interestingly, we can see that `set_maximum_version_registry!` handles an empty collection by other means:
+there is, in fact, no chance that `maximum(pkgversions)` will ever reach the error case.
+However, the compiler does not realize that, and therefore generates code that has been prepared to call `reduce_empty`, and that makes it vulnerable to invalidation.
+
+There are a couple of potential fixes.
+From here, we see that a potentially better definition `reduce_empty` for `VersionNumber`s might be
+
+```
+reduce_empty(::Base.BottomRF{typeof(max)}, ::Type{VersionNumber}) = VersionNumber(0)
+```
+
+In that case we could delete the `length` check and rely on this to produce the desired outcome.
+One should take this approach only if one can be convinced that version 0 can act as a universal "neutral element" for reductions involving `max` over `VersionNumber`s.
+Adding this to `Pkg` would be type-piracy, since `max`, `BottomRF`, and `VersionNumber` are all defined in `Base`, so fixing it this way would best be done in Base.
+
+But there are other possible fixes.
+A little higher up the tree we see a call to `mapreduce`, and this presents another opportunity because `mapreduce` allows you to supply an `init` value:
+
+```julia-repl
+julia> mapreduce(identity, max, Set(VersionNumber[]); init=VersionNumber(0))
+v"0.0.0"
+```
+
+Perhaps we could just call this instead of `maximum`.
+However, it's a bit uglier than the original;
+perhaps a nicer approach would be to allow one to supply `init` as a keyword argument to `maximum` itself.
+While this is not supported on Julia versions up through 1.5, it's a feature that seems to make sense, and this analysis suggests that it might also allow developers to make code more robust against certain kinds of invalidation.
```

Comment:
I admit to reading this (and various of the other suggestions for potentially removing invalidations) with a certain amount of dismay!

For a post on a general blog like the julialang blog, I feel like it sends an odd message to readers: "Be careful of abstractions you guys! Actually you might need to know about the implementation detail and call into that on occasion!"

Seasoned Julia developers will know that there's a maintenance and clarity cost and have an idea of when to make the tradeoff. But I fear that won't be true of everyone reading this post.

Reply:
Yeah, I had a similar comment https://github.com/JuliaLang/www.julialang.org/pull/794#discussion_r424454133

(Having said that, I think this particular part about `init` makes totally sense. `reduce` is not well-defined on possibly-empty collection if `op` is not a monoid; i.e., it needs `init` in general. So the invalidation hunt teaches us it's good to pay attention to algebra :smile:.)